File: packages/fullmag-py/src/fullmag/runtime/scene_document.py
from __future__ import annotations

from typing import Any


def _material_id(name: str) -> str:
    return f"mat:{name}"


def _magnetization_id(name: str) -> str:
    return f"mag:{name}"


def _zero_vec3() -> list[float]:
    return [0.0, 0.0, 0.0]


def _one_vec3() -> list[float]:
    return [1.0, 1.0, 1.0]


def _identity_quat() -> list[float]:
    return [0.0, 0.0, 0.0, 1.0]


def _default_mapping() -> dict[str, object]:
    return {
        "space": "object",
        "projection": "object_local",
        "clamp_mode": "none",
    }


def _default_texture_transform() -> dict[str, object]:
    return {
        "translation": _zero_vec3(),
        "rotation_quat": _identity_quat(),
        "scale": _one_vec3(),
        "pivot": _zero_vec3(),
    }


_INTERACTION_ORDER = (
    "exchange",
    "demag",
    "interfacial_dmi",
    "uniaxial_anisotropy",
)


def _normalize_axis3(value: object) -> list[float]:
    if isinstance(value, list) and len(value) == 3:
        try:
            return [float(value[0]), float(value[1]), float(value[2])]
        except (TypeError, ValueError):
            return [0.0, 0.0, 1.0]
    return [0.0, 0.0, 1.0]


def _default_interaction_params(kind: str, *, material_dind: object) -> dict[str, object] | None:
    if kind == "interfacial_dmi":
        dind = _number_or_none(material_dind)
        return {"dind": dind if dind is not None else 1e-3}
    if kind == "uniaxial_anisotropy":
        return {"ku1": 0.0, "axis": [0.0, 0.0, 1.0]}
    return None


def _normalize_interaction_entry(
    raw: object,
    *,
    material_dind: object,
) -> dict[str, object] | None:
    if not isinstance(raw, dict):
        return None
    kind = str(raw.get("kind") or "").strip()
    if kind not in _INTERACTION_ORDER:
        return None
    if kind in {"exchange", "demag"}:
        return {"kind": kind, "enabled": True, "params": None}
    params = raw.get("params")
    params_map = dict(params) if isinstance(params, dict) else (_default_interaction_params(kind, material_dind=material_dind) or {})
    if kind == "interfacial_dmi":
        dind = _number_or_none(params_map.get("dind"))
        if dind is None:
            dind = _number_or_none(material_dind)
        params_map["dind"] = dind if dind is not None else 1e-3
    elif kind == "uniaxial_anisotropy":
        ku1 = _number_or_none(params_map.get("ku1"))
        params_map["ku1"] = ku1 if ku1 is not None else 0.0
        params_map["axis"] = _normalize_axis3(params_map.get("axis"))
    return {
        "kind": kind,
        "enabled": bool(raw.get("enabled", True)),
        "params": params_map,
    }


def _ensure_physics_stack(raw: object, *, material_dind: object = None) -> list[dict[str, object]]:
    by_kind: dict[str, dict[str, object]] = {}
    if isinstance(raw, list):
        for entry in raw:
            normalized = _normalize_interaction_entry(entry, material_dind=material_dind)
            if normalized is not None:
                by_kind[str(normalized["kind"])] = normalized
    for required in ("exchange", "demag"):
        by_kind[required] = {"kind": required, "enabled": True, "params": None}
    if material_dind is not None and "interfacial_dmi" not in by_kind:
        by_kind["interfacial_dmi"] = _normalize_interaction_entry(
            {"kind": "interfacial_dmi", "enabled": True, "params": None},
            material_dind=material_dind,
        ) or {"kind": "interfacial_dmi", "enabled": True, "params": {"dind": 1e-3}}
    ordered: list[dict[str, object]] = []
    for kind in _INTERACTION_ORDER:
        entry = by_kind.get(kind)
        if entry is not None:
            ordered.append(entry)
    return ordered


def build_scene_document_from_builder(builder: dict[str, Any]) -> dict[str, Any]:
    geometries = builder.get("geometries") or []
    objects: list[dict[str, Any]] = []
    materials: list[dict[str, Any]] = []
    magnetization_assets: list[dict[str, Any]] = []

    for geometry in geometries:
        name = str(geometry.get("name", "object"))
        geometry_params = dict(geometry.get("geometry_params") or {})
        translation = geometry_params.pop("translation", geometry_params.pop("translate", [0, 0, 0]))
        magnetization = dict(geometry.get("magnetization") or {})
        mag_kind = str(magnetization.get("kind", "uniform"))
        if mag_kind == "file" and (
            magnetization.get("dataset") is not None or magnetization.get("sample_index") is not None
        ):
            mag_kind = "sampled"

        material_properties = dict(geometry.get("material") or {})
        physics_stack = _ensure_physics_stack(
            geometry.get("physics_stack"),
            material_dind=material_properties.get("Dind"),
        )

        objects.append(
            {
                "id": name,
                "name": name,
                "geometry": {
                    "geometry_kind": str(geometry.get("geometry_kind", "")),
                    "geometry_params": geometry_params,
                    "bounds_min": geometry.get("bounds_min"),
                    "bounds_max": geometry.get("bounds_max"),
                },
                "transform": {
                    "translation": translation if isinstance(translation, list) else [0, 0, 0],
                    "rotation_quat": _identity_quat(),
                    "scale": _one_vec3(),
                    "pivot": _zero_vec3(),
                },
                "material_ref": _material_id(name),
                "region_name": geometry.get("region_name"),
                "magnetization_ref": _magnetization_id(name),
                "physics_stack": physics_stack,
                "object_mesh": geometry.get("mesh"),
                "mesh_override": geometry.get("mesh"),
                "visible": True,
                "locked": False,
                "tags": [],
            }
        )
        materials.append(
            {
                "id": _material_id(name),
                "name": f"{name} material",
                "properties": material_properties,
            }
        )
        magnetization_assets.append(
            {
                "id": _magnetization_id(name),
                "name": f"{name} magnetization",
                "kind": mag_kind,
                "value": magnetization.get("value"),
                "seed": magnetization.get("seed"),
                "source_path": magnetization.get("source_path"),
                "source_format": magnetization.get("source_format"),
                "dataset": magnetization.get("dataset"),
                "sample_index": magnetization.get("sample_index"),
                "mapping": dict(magnetization.get("mapping") or _default_mapping()),
                "texture_transform": dict(
                    magnetization.get("texture_transform") or _default_texture_transform()
                ),
                "preset_kind": magnetization.get("preset_kind"),
                "preset_params": magnetization.get("preset_params"),
                "preset_version": magnetization.get("preset_version"),
                "ui_label": magnetization.get("ui_label"),
            }
        )

    return {
        "version": "scene.v1",
        "revision": int(builder.get("revision", 0)),
        "scene": {
            "id": "scene",
            "name": "Scene",
            "source_of_truth": "repo_head",
            "authoring_schema": "mesh-first-fem.v1",
        },
        "universe": builder.get("universe"),
        "objects": objects,
        "materials": materials,
        "magnetization_assets": magnetization_assets,
        "current_modules": {
            "modules": builder.get("current_modules") or [],
            "excitation_analysis": builder.get("excitation_analysis"),
        },
        "study": {
            "backend": builder.get("backend"),
            "demag_realization": builder.get("demag_realization"),
            "external_field": builder.get("external_field"),
            "solver": builder.get("solver") or {},
            "universe_mesh": builder.get("universe"),
            "shared_domain_mesh": builder.get("mesh") or {},
            "mesh_defaults": builder.get("mesh") or {},
            "stages": builder.get("stages") or [],
            "study_pipeline": builder.get("study_pipeline"),
            "initial_state": builder.get("initial_state"),
        },
        "outputs": {"items": []},
        "editor": {
            "selected_object_id": None,
            "gizmo_mode": None,
            "transform_space": None,
            "selected_entity_id": None,
            "focused_entity_id": None,
            "object_view_mode": "context",
            "air_mesh_visible": True,
            "air_mesh_opacity": 28.0,
            "mesh_entity_view_state": {},
            "active_transform_scope": None,
        },
    }


def build_builder_from_scene_document(scene: dict[str, Any]) -> dict[str, Any]:
    materials = {
        str(material.get("id", "")): dict(material.get("properties") or {})
        for material in (scene.get("materials") or [])
    }
    magnetization_assets = {
        str(asset.get("id", "")): dict(asset)
        for asset in (scene.get("magnetization_assets") or [])
    }
    geometries: list[dict[str, Any]] = []

    for obj in scene.get("objects") or []:
        material_ref = str(obj.get("material_ref") or "")
        if not material_ref or material_ref not in materials:
            raise ValueError(
                f"object '{obj.get('id') or obj.get('name') or ''}' references missing material '{material_ref}'"
            )
        magnetization_ref = str(obj.get("magnetization_ref") or "")
        if not magnetization_ref or magnetization_ref not in magnetization_assets:
            raise ValueError(
                f"object '{obj.get('id') or obj.get('name') or ''}' references missing magnetization asset '{magnetization_ref}'"
            )
        geometry = dict(obj.get("geometry") or {})
        geometry_params = dict(geometry.get("geometry_params") or {})
        transform = dict(obj.get("transform") or {})
        translation = transform.get("translation")
        if isinstance(translation, list) and len(translation) == 3:
            if any(abs(float(value)) > 0 for value in translation):
                geometry_params["translation"] = [float(value) for value in translation]

        magnetization_asset = magnetization_assets[magnetization_ref]
        magnetization = {
            "kind": str(magnetization_asset.get("kind", "uniform")),
            "value": magnetization_asset.get("value"),
            "seed": magnetization_asset.get("seed"),
            "source_path": magnetization_asset.get("source_path"),
            "source_format": magnetization_asset.get("source_format"),
            "dataset": magnetization_asset.get("dataset"),
            "sample_index": magnetization_asset.get("sample_index"),
            "mapping": dict(magnetization_asset.get("mapping") or _default_mapping()),
            "texture_transform": dict(
                magnetization_asset.get("texture_transform") or _default_texture_transform()
            ),
            "preset_kind": magnetization_asset.get("preset_kind"),
            "preset_params": magnetization_asset.get("preset_params"),
            "preset_version": magnetization_asset.get("preset_version"),
            "ui_label": magnetization_asset.get("ui_label"),
        }
        material_properties = materials[material_ref]
        physics_stack = _ensure_physics_stack(
            obj.get("physics_stack"),
            material_dind=material_properties.get("Dind"),
        )

        geometries.append(
            {
                "name": str(obj.get("name") or obj.get("id") or ""),
                "region_name": obj.get("region_name"),
                "geometry_kind": str(geometry.get("geometry_kind", "")),
                "geometry_params": geometry_params,
                "bounds_min": geometry.get("bounds_min"),
                "bounds_max": geometry.get("bounds_max"),
                "material": material_properties,
                "magnetization": magnetization,
                "physics_stack": physics_stack,
                "mesh": obj.get("object_mesh", obj.get("mesh_override")),
            }
        )

    study = dict(scene.get("study") or {})
    current_modules = dict(scene.get("current_modules") or {})
    return {
        "revision": int(scene.get("revision", 0)),
        "backend": study.get("backend"),
        "demag_realization": study.get("demag_realization"),
        "external_field": study.get("external_field"),
        "solver": study.get("solver") or {},
        "mesh": study.get("shared_domain_mesh") or study.get("mesh_defaults") or {},
        "universe": study.get("universe_mesh") or scene.get("universe"),
        "stages": study.get("stages") or [],
        "study_pipeline": study.get("study_pipeline"),
        "initial_state": study.get("initial_state"),
        "geometries": geometries,
        "current_modules": current_modules.get("modules") or [],
        "excitation_analysis": current_modules.get("excitation_analysis"),
    }


def builder_overrides_from_scene_document(scene: dict[str, Any]) -> dict[str, Any]:
    builder = build_builder_from_scene_document(scene)
    solver = dict(builder.get("solver") or {})
    mesh = dict(builder.get("mesh") or {})
    return {
        "demag_realization": builder.get("demag_realization"),
        "external_field": (
            [float(value) for value in builder.get("external_field")]
            if isinstance(builder.get("external_field"), list)
            and len(builder.get("external_field")) == 3
            else None
        ),
        "solver": {
            "integrator": solver.get("integrator") or None,
            "fixed_timestep": _number_or_none(solver.get("fixed_timestep")),
            "relax": {
                "algorithm": solver.get("relax_algorithm") or None,
                "torque_tolerance": _number_or_none(solver.get("torque_tolerance")),
                "energy_tolerance": _number_or_none(solver.get("energy_tolerance")),
                "max_steps": _int_or_none(solver.get("max_relax_steps")),
            },
        },
        "mesh": {
            "algorithm_2d": mesh.get("algorithm_2d"),
            "algorithm_3d": mesh.get("algorithm_3d"),
            "hmax": _number_or_auto(mesh.get("hmax")),
            "hmin": _number_or_none(mesh.get("hmin")),
            "size_factor": mesh.get("size_factor"),
            "size_from_curvature": mesh.get("size_from_curvature"),
            "growth_rate": _number_or_none(mesh.get("growth_rate")),
            "narrow_regions": mesh.get("narrow_regions"),
            "smoothing_steps": mesh.get("smoothing_steps"),
            "optimize": mesh.get("optimize") or None,
            "optimize_iterations": mesh.get("optimize_iterations"),
            "compute_quality": mesh.get("compute_quality"),
            "per_element_quality": mesh.get("per_element_quality"),
            "adaptive_mesh": None
            if not mesh.get("adaptive_enabled")
            else {
                "enabled": True,
                "policy": mesh.get("adaptive_policy"),
                "theta": mesh.get("adaptive_theta"),
                "h_min": _number_or_none(mesh.get("adaptive_h_min")),
                "h_max": _number_or_none(mesh.get("adaptive_h_max")),
                "max_passes": mesh.get("adaptive_max_passes"),
                "error_tolerance": _number_or_none(mesh.get("adaptive_error_tolerance")),
            },
        },
        "universe": builder.get("universe"),
        "stages": [
            {
                "kind": stage.get("kind"),
                "entrypoint_kind": stage.get("entrypoint_kind"),
                "integrator": stage.get("integrator") or None,
                "fixed_timestep": _number_or_none(stage.get("fixed_timestep")),
                "until_seconds": _number_or_none(stage.get("until_seconds")),
                "relax_algorithm": stage.get("relax_algorithm") or None,
                "torque_tolerance": _number_or_none(stage.get("torque_tolerance")),
                "energy_tolerance": _number_or_none(stage.get("energy_tolerance")),
                "max_steps": _int_or_none(stage.get("max_steps")),
                "eigen_count": _int_or_none(stage.get("eigen_count")),
                "eigen_target": stage.get("eigen_target") or None,
                "eigen_include_demag": (
                    bool(stage.get("eigen_include_demag"))
                    if isinstance(stage.get("eigen_include_demag"), bool)
                    else None
                ),
                "eigen_equilibrium_source": stage.get("eigen_equilibrium_source") or None,
                "eigen_normalization": stage.get("eigen_normalization") or None,
                "eigen_target_frequency": _number_or_none(stage.get("eigen_target_frequency")),
                "eigen_damping_policy": stage.get("eigen_damping_policy") or None,
                "eigen_k_vector": stage.get("eigen_k_vector") or None,
                "eigen_spin_wave_bc": stage.get("eigen_spin_wave_bc") or None,
                "eigen_spin_wave_bc_config": (
                    dict(stage.get("eigen_spin_wave_bc_config"))
                    if isinstance(stage.get("eigen_spin_wave_bc_config"), dict)
                    else None
                ),
            }
            for stage in (builder.get("stages") or [])
        ],
        "study_pipeline": builder.get("study_pipeline"),
        "initial_state": builder.get("initial_state"),
        "geometries": builder.get("geometries") or [],
        "current_modules": builder.get("current_modules") or [],
        "excitation_analysis": builder.get("excitation_analysis"),
    }


def _number_or_none(value: Any) -> float | str | None:
    if value is None:
        return None
    if isinstance(value, str):
        stripped = value.strip()
        if not stripped:
            return None
        if stripped.lower() == "auto":
            return "auto"
        try:
            return float(stripped)
        except ValueError:
            return None
    if isinstance(value, (int, float)):
        return float(value)
    return None


def _number_or_auto(value: Any) -> float | str | None:
    return _number_or_none(value)


def _int_or_none(value: Any) -> int | None:
    if value is None:
        return None
    if isinstance(value, str):
        stripped = value.strip()
        if not stripped:
            return None
        try:
            return int(stripped)
        except ValueError:
            return None
    if isinstance(value, (int, float)):
        return int(value)
    return None

File: apps/web/lib/session/sceneDocument.ts
import type {
  MagnetizationAsset,
  SceneDocument,
  SceneObject,
  ScriptBuilderGeometryEntry,
  ScriptBuilderMagnetizationEntry,
  ScriptBuilderState,
  Transform3D,
} from "./types";
import { ensureObjectPhysicsStack } from "./magneticPhysics";

function zeroVec3(): [number, number, number] {
  return [0, 0, 0];
}

function oneVec3(): [number, number, number] {
  return [1, 1, 1];
}

function identityQuat(): [number, number, number, number] {
  return [0, 0, 0, 1];
}

function cloneGeometryParams(
  value: Record<string, unknown> | null | undefined,
): Record<string, unknown> {
  return value ? { ...value } : {};
}

function readTranslation(params: Record<string, unknown>): [number, number, number] {
  const raw =
    (Array.isArray(params.translation) ? params.translation : null) ??
    (Array.isArray(params.translate) ? params.translate : null);
  if (!raw || raw.length !== 3) {
    return zeroVec3();
  }
  return [Number(raw[0] ?? 0), Number(raw[1] ?? 0), Number(raw[2] ?? 0)];
}

function stripTranslation(params: Record<string, unknown>): Record<string, unknown> {
  const next = { ...params };
  delete next.translation;
  delete next.translate;
  return next;
}

function materialIdForGeometry(name: string): string {
  return `mat:${name}`;
}

function magnetizationIdForGeometry(name: string): string {
  return `mag:${name}`;
}

function identityTransformWithTranslation(
  translation: [number, number, number],
): Transform3D {
  return {
    translation,
    rotation_quat: identityQuat(),
    scale: oneVec3(),
    pivot: zeroVec3(),
  };
}

function buildMagnetizationAsset(
  name: string,
  magnetization: ScriptBuilderMagnetizationEntry,
): MagnetizationAsset {
  const inferredKind =
    magnetization.kind === "file" &&
    (magnetization.dataset != null || magnetization.sample_index != null)
      ? "sampled"
      : magnetization.kind;
  return {
    id: magnetizationIdForGeometry(name),
    name: `${name} magnetization`,
    kind: inferredKind,
    value: magnetization.value ?? null,
    seed: magnetization.seed ?? null,
    source_path: magnetization.source_path ?? null,
    source_format: magnetization.source_format ?? null,
    dataset: magnetization.dataset ?? null,
    sample_index: magnetization.sample_index ?? null,
    mapping: magnetization.mapping ?? {
      space: "object",
      projection: "object_local",
      clamp_mode: "clamp",
    },
    texture_transform: magnetization.texture_transform ?? {
      translation: zeroVec3(),
      rotation_quat: identityQuat(),
      scale: oneVec3(),
      pivot: zeroVec3(),
    },
    preset_kind: magnetization.preset_kind ?? null,
    preset_params: magnetization.preset_params ?? null,
    preset_version: magnetization.preset_version ?? null,
    ui_label: magnetization.ui_label ?? null,
  };
}

export function buildSceneDocumentFromScriptBuilder(
  builder: ScriptBuilderState,
): SceneDocument {
  const objects: SceneObject[] = builder.geometries.map((geometry) => {
    const geometryParams = cloneGeometryParams(geometry.geometry_params);
    const translation = readTranslation(geometryParams);
    return {
      id: geometry.name,
      name: geometry.name,
      geometry: {
        geometry_kind: geometry.geometry_kind,
        geometry_params: stripTranslation(geometryParams),
        bounds_min: geometry.bounds_min ?? null,
        bounds_max: geometry.bounds_max ?? null,
      },
      transform: identityTransformWithTranslation(translation),
      material_ref: materialIdForGeometry(geometry.name),
      region_name: geometry.region_name ?? null,
      magnetization_ref: magnetizationIdForGeometry(geometry.name),
      physics_stack: ensureObjectPhysicsStack(
        geometry.physics_stack,
        geometry.material.Dind,
      ),
      object_mesh: geometry.mesh ?? null,
      mesh_override: geometry.mesh ?? null,
      visible: true,
      locked: false,
      tags: [],
    };
  });

  return {
    version: "scene.v1",
    revision: builder.revision,
    scene: {
      id: "scene",
      name: "Scene",
      source_of_truth: "repo_head",
      authoring_schema: "mesh-first-fem.v1",
    },
    universe: builder.universe,
    objects,
    materials: builder.geometries.map((geometry) => ({
      id: materialIdForGeometry(geometry.name),
      name: `${geometry.name} material`,
      properties: geometry.material,
    })),
    magnetization_assets: builder.geometries.map((geometry) =>
      buildMagnetizationAsset(geometry.name, geometry.magnetization),
    ),
    current_modules: {
      modules: builder.current_modules,
      excitation_analysis: builder.excitation_analysis,
    },
    study: {
      backend: builder.backend,
      requested_backend: builder.backend ?? "auto",
      requested_device: "auto",
      requested_precision: "double",
      requested_mode: "strict",
      demag_realization: builder.demag_realization,
      external_field: builder.external_field,
      solver: builder.solver,
      universe_mesh: builder.universe,
      shared_domain_mesh: builder.mesh,
      mesh_defaults: builder.mesh,
      stages: builder.stages,
      study_pipeline: builder.study_pipeline ?? null,
      initial_state: builder.initial_state,
    },
    outputs: { items: [] },
    editor: {
      selected_object_id: null,
      gizmo_mode: null,
      transform_space: null,
      selected_entity_id: null,
      focused_entity_id: null,
      object_view_mode: "context",
      air_mesh_visible: false,
      air_mesh_opacity: 28,
      mesh_entity_view_state: {},
      visualization_presets: [],
      active_visualization_preset_ref: null,
      active_transform_scope: null,
    },
  };
}

function magnetizationForObject(
  scene: SceneDocument,
  object: SceneObject,
): ScriptBuilderMagnetizationEntry {
  const asset = scene.magnetization_assets.find(
    (candidate) => candidate.id === object.magnetization_ref,
  );
  if (!asset) {
    return {
      kind: "uniform",
      value: [1, 0, 0],
      seed: null,
      source_path: null,
      source_format: null,
      dataset: null,
      sample_index: null,
    };
  }
  return {
    kind: asset.kind,
    value: asset.value,
    seed: asset.seed,
    source_path: asset.source_path,
    source_format: asset.source_format,
    dataset: asset.dataset,
    sample_index: asset.sample_index,
    mapping: asset.mapping,
    texture_transform: asset.texture_transform,
    preset_kind: asset.preset_kind,
    preset_params: asset.preset_params,
    preset_version: asset.preset_version,
    ui_label: asset.ui_label,
  };
}

export function buildScriptBuilderFromSceneDocument(
  scene: SceneDocument,
): ScriptBuilderState {
  return {
    revision: scene.revision,
    backend: scene.study.backend,
    demag_realization: scene.study.demag_realization,
    external_field: scene.study.external_field,
    solver: scene.study.solver,
    mesh: scene.study.shared_domain_mesh ?? scene.study.mesh_defaults,
    universe: scene.study.universe_mesh ?? scene.universe,
    domain_frame: null,
    stages: scene.study.stages,
    study_pipeline: scene.study.study_pipeline ?? null,
    initial_state: scene.study.initial_state,
    geometries: scene.objects.map((object): ScriptBuilderGeometryEntry => {
      const material =
        scene.materials.find((candidate) => candidate.id === object.material_ref)?.properties ?? {
          Ms: null,
          Aex: null,
          alpha: 0.01,
          Dind: null,
        };
      const geometryParams = stripTranslation(
        cloneGeometryParams(object.geometry.geometry_params),
      );
      const translation = object.transform.translation;
      if (translation.some((value) => Math.abs(value) > Number.EPSILON)) {
        geometryParams.translation = [...translation];
      }
      return {
        name: object.name || object.id,
        region_name: object.region_name ?? null,
        geometry_kind: object.geometry.geometry_kind,
        geometry_params: geometryParams,
        bounds_min: object.geometry.bounds_min ?? null,
        bounds_max: object.geometry.bounds_max ?? null,
        material,
        physics_stack: ensureObjectPhysicsStack(
          object.physics_stack,
          material.Dind,
        ),
        magnetization: magnetizationForObject(scene, object),
        mesh: object.object_mesh ?? object.mesh_override ?? null,
      };
    }),
    current_modules: scene.current_modules.modules,
    excitation_analysis: scene.current_modules.excitation_analysis,
  };
}

export function sceneDocumentSignature(scene: SceneDocument | null): string | null {
  return scene ? JSON.stringify(scene) : null;
}
